src/graphs.rs

```diff
@@ -0,0 +1,312 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+
+use std::collections::BTreeMap;
+use std::collections::BTreeSet;
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::collections::VecDeque;
+
+use crate::NpmPackageInfo;
+use crate::PackagesContent;
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+enum LockfilePkgId {
+  Npm(LockfileNpmPackageId),
+  Jsr(LockfileJsrPkgNv),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+struct LockfileJsrPkgNv(String);
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+struct LockfileNpmPackageId(String);
+
+impl LockfileNpmPackageId {
+  pub fn parts(&self) -> impl Iterator<Item = &str> {
+    let package_id = &self.0;
+    let package_id = package_id.strip_prefix("npm:").unwrap_or(package_id);
+    package_id.split('_').filter(|s| !s.is_empty())
+  }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+struct LockfilePkgReq(String);
+
+enum LockfileGraphPackage {
+  Jsr(LockfileJsrGraphPackage),
+  Npm(LockfileNpmGraphPackage),
+}
+
+struct LockfileNpmGraphPackage {
+  /// Root ids that transitively reference this package.
+  root_ids: HashSet<LockfilePkgId>,
+  integrity: String,
+  dependencies: BTreeMap<String, LockfileNpmPackageId>,
+}
+
+#[derive(Default)]
+struct LockfileJsrGraphPackage {
+  /// Root ids that transitively reference this package.
+  root_ids: HashSet<LockfilePkgId>,
+  dependencies: BTreeSet<LockfilePkgReq>,
+}
+
+/// Graph used to analyze a lockfile to determine which packages
+/// and remotes can be removed based on config file changes.
+pub struct LockfilePackageGraph<FNvToJsrUrl: Fn(&str) -> Option<String>> {
+  root_packages: HashMap<LockfilePkgReq, LockfilePkgId>,
+  packages: HashMap<LockfilePkgId, LockfileGraphPackage>,
+  remotes: BTreeMap<String, String>,
+  nv_to_jsr_url: FNvToJsrUrl,
+}
+
+impl<FNvToJsrUrl: Fn(&str) -> Option<String>>
+  LockfilePackageGraph<FNvToJsrUrl>
+{
+  pub fn from_lockfile<'a>(
+    content: PackagesContent,
+    remotes: BTreeMap<String, String>,
+    old_config_file_packages: impl Iterator<Item = &'a str>,
+    nv_to_jsr_url: FNvToJsrUrl,
+  ) -> Self {
+    let mut root_packages =
+      HashMap::<LockfilePkgReq, LockfilePkgId>::with_capacity(
+        content.specifiers.len(),
+      );
+    // collect the specifiers to version mappings
+    let mut packages = HashMap::new();
```

Comment:
Could you pre-allocate this? I think it's better to over-allocate than having to resize the hash map